[1.8] Add additional tests for object-based APIs

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51341 [1.8] Add additional tests for object-based APIs**
* #50625 [Collective APIs] Make python object collective API args consistent

Adds tests for objects that contain CPU/GPU tensors to ensure that
they can also be serialized/deserialized appropriately.

Differential Revision: [D26144100](https://our.internmc.facebook.com/intern/diff/D26144100/)

rohan-varma committed:
[1.8] Add additional tests for object-based APIs

Adds tests for objects that contain CPU/GPU tensors to ensure that
they can also be serialized/deserialized appropriately.

Differential Revision: [D26144100](https://our.internmc.facebook.com/intern/diff/D26144100/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[1.8] Add additional tests for object-based APIs"


Adds tests for objects that contain CPU/GPU tensors to ensure that
they can also be serialized/deserialized appropriately.

Differential Revision: [D26144100](https://our.internmc.facebook.com/intern/diff/D26144100/)

[ghstack-poisoned]

SciPioneer commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -57,14 +57,27 @@ def __init__(self, x):
 
     def __eq__(self, other):
-        return self.__dict__ == other.__dict__
+        def eq(value, other):
```
Probably better to add a comment on `Foo` class, mentioning that x can be either a tensor or an object.

rohan-varma committed:
Update on "[1.8] Add additional tests for object-based APIs"


Adds tests for objects that contain CPU/GPU tensors to ensure that
they can also be serialized/deserialized appropriately.

Differential Revision: [D26144100](https://our.internmc.facebook.com/intern/diff/D26144100/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[1.8] Add additional tests for object-based APIs"


Adds tests for objects that contain CPU/GPU tensors to ensure that
they can also be serialized/deserialized appropriately.

Differential Revision: [D26144100](https://our.internmc.facebook.com/intern/diff/D26144100/)

[ghstack-poisoned]
